artsy/force
# chore: EM-1355 - send address verified by

The type of this PR is: **Chore**

cc @artsy/emerald-devs 
This PR solves [EM-1295] and mostly completes  [EM-1355]. It includes an update of the changes in https://github.com/artsy/force/pull/12636

### Description

With this change the `address_verified_by` argument is sent to [exchange's `setShipping` mutation](https://github.com/artsy/exchange/blob/b9079eb9d9bb5b157fb55f8d67b950e9f3ab2a73/app/graphql/mutations/set_shipping.rb#L30), triggering the backend address verification job.

It also fixes some types and cleans up some development code.


To accomplish this I updated our boolean `addressHasBeenVerified` state variable to be a `null | "ARTSY" | "USER"` `addressVerifiedBy` variable - its presence means that it has been verified, and editing the form after saving sets the address as verified by the user (since we don't perform verification more than once without a reload).

**Note/followup:** After the address has a verified-by value in exchange, it [will not be re-set](https://github.com/artsy/exchange/blob/b9079eb9d9bb5b157fb55f8d67b950e9f3ab2a73/app/services/shipping/verify_address.rb#L6). Thus if a user goes back and edits their address neither this nor the payload will be updated. Fixed in https://github.com/artsy/exchange/pull/1735

### Steps to test manually
1. Have an exchange staging console open
2. Begin an order checkout, removing all saved addresses
3. Enter a completely wrong US address for the delivery address and see the address verification modal - the version with no choices.
4. Click to keep the address
5. Click save & continue: 

**Find the order in the exchange staging console, see that the shipping address has been set along with `shipping_address_verified_by: "USER"` but no `shipping_address_verified_payload`.**

6. Reload the page
7. Enter a new address that will get suggestions, (401 broadway, 26th floor, new york, NY 10013 with the floor number on line 2 usually works).
8. Select the suggested address ['use this address']
9. Click save & continue again

**Reload the order in the exchange staging console, see that the shipping address has been updated along with `shipping_address_verified_by: "ARTSY"` and a `shipping_address_verified_payload`.**

10. On the shipping page with an address that has already been saved as verified by ARTSY (e.g. an Arta order that has not had its shipping option selected yet), edit the form to change something and save it again.

**Reload the order in exchange once more and see that the address has updated and is again verified by USER.**

Other expected behavior:
- User closes either modal without actively clicking the primary address select button: `address_verified_by USER`
- address verification not enabled: `address_verified_by: nil`

<details><summary>Saved address verification payload in exchange for an ARTSY-verified address</summary>

![image](https://github.com/artsy/force/assets/9088720/76be301a-5d16-4aa9-a88c-16f355f33cae)

</details>
